### ISSUE TITLE
UX requirements around switching to PyQt6 by default.

### ISSUE THREAD

**Issue by toofar** [qutebrowser/qutebrowser#7656]
UX requirements around switching to PyQt6 by default.

We are going to be switching the default Qt backend across a major version soon, from Qt5 to Qt6. The plan is that the backend, if not manually provided, will be autoselected. The current method for autoselection is that we try to import PyQt6 and if that fails try to import PyQt5.

This issue is to describe some use cases and discuss the importance of various improvements we could make. Possibly it would be better as a discussion so the individual points could be discussed in their own threads.

I had a bunch of repetitive user stories but it wasn't very interesting and I don't think anyone wants to read all them. (Eg I've updated my distro, the packager switched the dependencies over, I still have Qt5 installed.) So I boiled them down to some bullet points, with examples in sublists:

Variables:
1. whether a wrapper is forced or autoselection is used
2. whether you have Qt5 installed alongside Qt6 or not
    * currently we don't look at this at any time apart from autoselection. But as a user it defines what is possible for you.

Upgrade paths:
1. autoselection and you now only have PyQt6 available
    * for example your OS's packager changed the dependancies for the qute package and you didn't have any other applications that depended on Qt5
    * or you've set up a new virtualenv and are using that with an existing basedir
2. autoselection and you have both PyQt5 and 6 available
    * probably the main upgrade case for distro users
3. forced wrapper version
    * maybe you or your distro wanted to remove ambiguity by forcing the wrapper version
    * you have two qutebrowser versions installed and launched the newer one on the older basedir
    * I'm not sure there is a good use case for forcing the backend as a means of upgrading currently as we upgrade by default anyway

Downgrade paths:
1. forced wrapper version
    * there was some issue with Qt6 and you want to go back to Qt5
    * you have a new and old qutebrowser package installed and you launched the lower version one on the newer basedir
2. autoselection and you no longer have PyQt6 available
    * could happen if your Qt6 install is broken but you still have PyQt5 installed, or if you are using the same basedir across different virtualenvs

Questions:
1. are any packagers expecting to force the Qt wrapper version?
2. should we still show a warning when it's likely an OS packager has arranged things so that a user will be upgraded to a certain set of dependancies?
3. do we want to surface the wrapper selection logic (force or auto) in :version?
4. is there any use case that'll be enabled by storing the last used wrapper in the state file? Eg if you have both 5 and 6 installed and forced Qt5 do we remember that and not do autoselection until you change to forcing Qt6?
5. do we need to put more info in the upgrade warning like "because we detected a newer Qt version" or "because a different Qt version was selected by $mechanism"?
6. in the upgrade warning, when you have both versions installed, should we add a "continue with $other_version" button?
7. is the current autoselection logic comprehensive enough? Currently we only try to import the top level packages (PyQt5/6), if something like importing WebEngine or QtNetwork later fails we wouldn't catch that till later. Is that something that is worth accounting for?
8. can we add in a CLI argument for selecting the wrapper? Using an env var is a bit more cumbersome currently.

Current warnings: https://github.com/qutebrowser/qutebrowser/blob/7ab16cc/qutebrowser/misc/backendproblem.py#L335

1. Major version update warning. [Ok] [Abort]
    
![image](https://user-images.githubusercontent.com/7419144/230846710-8f092d7c-483c-4364-8738-0c43af1df921.png)

2. Downgrade detected warning. [Ok] [Abort]
    
![image](https://user-images.githubusercontent.com/7419144/230846579-8f1fe65c-6c15-43c6-b3ae-84e3df3ed7b7.png)

3. Chosen wrapper not available error. [Ok]
    
![image](https://user-images.githubusercontent.com/7419144/230846824-0d92fefe-b6de-429b-afd4-cd3c2e6c5d1a.png)

4. No wrapper found.

Stack trace from error message `qutebrowser.qt.machinery.Error: No Qt wrapper found, tried PyQt6, PyQt5`

**Comment by The-Compiler** [comment]
Thanks for the excellent write-up, much appreciated!

One thing I think we should do is [just catch `ModuleNotFoundError`](https://docs.python.org/3/library/exceptions.html#ModuleNotFoundError) (rather than `ImportError`), and have the latter open some error dialog instead.

I'll try to take a look at the rest of those later this week. Seems like the last big roadblock before flipping the switch, unless I'm missing something.

**Comment by ionenwks** [comment]
> 1. are any packagers expecting to force the Qt wrapper version?

By force do you mean change the default? Already doing this in Gentoo (for the live/git test ebuild) with a machinery.py sed. To handle dependencies the user is expected to pick which they want at install time (much like other packages with --enable/disable-feature build time switches). Aka if have USE=qt6 then it'll pull the matching PyQt6 & friends and qute will use that by default (USE=-qt6 would opt-out and pull+use PyQt5).

It'd be unexpected if the user-picked backend wouldn't get used because the other PyQt* happened to be installed for other reasons and was auto-selected.

Imagine binary distros wouldn't offer the choice and just hard-pull PyQt6 dependency aligning with the the default here though, Gentoo packaging doesn't really lock users into anything within reason.

**Comment by toofar** [comment]
> One thing I think we should do is just catch ModuleNotFoundError (rather than ImportError), and have the latter open some error dialog instead.

Hmm, I'm not following. Change the [autoselect catch](https://github.com/qutebrowser/qutebrowser/blob/de82e1a/qutebrowser/qt/machinery.py#L43) to be a ModuleNotFoundError I presume. But catch ImportErrors from where? Everywhere? The same place (when would that happen)?

Here is my opinions:

1. are any packagers expecting to force the Qt wrapper version?
    * apparently so! I think it only makes sense when you are putting the choice on the user in an OS native way, like Gentoo's use flags apparently, or having multiple packages in Arch.
2. should we still show a warning when it's likely an OS packager has arranged
   things so that a user will be upgraded to a certain set of dependencies?
    * I think so. It's a fairly painless one-time click through of a dialog. And maintaining more logic to let the dialog be suppressed sounds like more work. The counter is that if a packager has a finely curated user experience and they don't think it's needed they may want to suppress it. But until someone makes a strong case for that I don't think its worth the effort.
3. do we want to surface the wrapper selection logic (force or auto) in :version?
    * yes
4. is there any use case that'll be enabled by storing the last used wrapper in the state file? Eg if you have both 5 and 6 installed and forced Qt5 do we remember that and not do autoselection until you change to forcing Qt6?
    * I'm not sure. The use cases I can think of for this are a) make the wrapper selection "sticky" so you don't have to provide the env var every time; or b) provide a warning when the wrapper changes; or c) prevent downgrades via autoselection.
    * For (a) we currently only initialize the state file after reading all the config. So we can't use the state file to make it transparently sticky, we could use to to show a warning + restart-with-override option. But in the case where you would want the selection to be sticky that's probably not a great UX?
    * For (b) the wrapper changing is currently a one-to-one match to major Qt version changes which we already have a warning for. So it's redundant right now. Would it be useful in the future when we have multiple wrappers for the same Qt version? I don't know. If you had our preferred wrapper installed and working and we switched to it would it have any user relevant effects? Not enough data.
    * For (c) putting more barriers in place before a downgrade might be something we want to explore. But again, for now its the same case as a major version downgrade which we already have covered.
    * TL;DR: we can store it but I'm not sure what we would use it for.
5. do we need to put more info in the upgrade warning like "because we detected a newer Qt version" or "because a different Qt version was selected by $mechanism"?
    * I guess it wouldn't hurt to put the wrapper override mechanism there if it was used. Although thinking about that opens up a can of worms in my lazy brain. Why not list all the available wrappers in there too? And tell people how to change it! And link to some docs! So much work.
6. in the upgrade warning, when you have both versions installed, should we add a "continue with $other_version" button?
    * We have this for the backend warnings but I'm not sure it makes sense here.
    * For upgrades I guess this would be like "I don't want to upgrade just yet, ask me again next time". We could add it via restart + override but since there aren't any major issues with Qt6 and we are going to recommend people upgrade then I'm not sure it's worth the effort.
    * For downgrades you should only be downgrading if you've specified the wrapper explicitly so I don't think it makes much sense there.
7. is the current autoselection logic comprehensive enough? Currently we only try to import the top level packages (PyQt5/6), if something like importing WebEngine or QtNetwork later fails we wouldn't catch that till later. Is that something that is worth accounting for?
    * :shrug: I have no opinion about this. It's too hypothetical for me.
8. can we add in a CLI argument for selecting the wrapper? Using an env var is a bit more cumbersome currently.
    * Yes we can, parsing the CLI args is the one thing we do before importing Qt.
    * What would the arg be? `--qt-wrapper` I guess.

Here's my proposed actions. Bit of a shorter list than I was expecting. Most of
the more complex changes we could make I don't think are worthwhile since
upgrading is uncomplicated (no-one's had any issues with it so far right?) and
we are recommending everyone does so. I may be missing problematic cases that
we should account for though!

* add whether the wrapper was explicitly selected or not to :version
* maybe add whether the wrapper was explicitly selected or not into the up and downgrade warning dialogs
* add debug logging in `_autoselect_wrapper()` ;)
* add a CLI arg to set the wrapper to make that a bit more discoverable (although that's mostly useful for developers switching between versions a bunch so probably not a huge deal really)

**Comment by The-Compiler** [comment]
I took a stab at (almost) all of this with #7741. Some thoughts and screenshots with that PR in below:

(sorry for the blurry screenshots, different DPI display as usual, maybe my screenshot tool got confused or something...)

## Packagers overriding Qt version

> > are any packagers expecting to force the Qt wrapper version?

> apparently so! I think it only makes sense when you are putting the choice on the user in an OS native way, like Gentoo's use flags apparently, or having multiple packages in Arch.

Decided to leave this out for the follow-up PR actually flipping the switch.

For Archlinux, I expect to merge the qutebrowser-qt6-git package into
qutebrowser-git once the flip is switched and add PyQt6 as dependency.
Someone in theory could upload a qutebrowser-qt5-git after that, but that's
probably not going to be me unless there is actual demand for it.

This pretty much leaves us Gentoo. Unfortunately, the `_DEFAULT_WRAPPER =
"PyQt5"` (as well as the given `sed` instructions) don't really make sense
anymore after autodetection is on.

My proposal here when flipping the switch would be to change this to
`_WRAPPER_OVERRIDE = None` which can be set by distributions (by patching the
source), and then if that is set, use that instead of autodetection.
I think `QUTE_QT_WRAPPER` and `--qt-wrapper` should still take precedence
though.

Unfortunately this means distributions will adjust their patching.
The naming/semantics of `_DEFAULT_WRAPPER` were a bit short-sighted in hindsight.

@ionenwks Any thoughts on this?

## Upgrading Qt

With #7741:

![image](https://github.com/qutebrowser/qutebrowser/assets/625793/ba20dc20-6f11-4762-b510-748eba67ac72)

> > should we still show a warning when it's likely an OS packager has arranged
> > things so that a user will be upgraded to a certain set of dependencies?

> I think so. It's a fairly painless one-time click through of a dialog. And maintaining more logic to let the dialog be suppressed sounds like more work. The counter is that if a packager has a finely curated user experience and they don't think it's needed they may want to suppress it. But until someone makes a strong case for that I don't think its worth the effort.

Agreed. It seems pretty much impossible for us to reliably know how the upgrade happened anyways. And IMHO, this gives the power to the user on what to do. If they have an important presentation coming up, they might want to cancel and run with `--qt-wrapper PyQt5`. If they care a lot about their browser data, they might want to take a backup before clicking ok. It's a big upgrade (arguably the biggest one since QtWebKit -> QtWebEngine), so I think it's good to give the user info about what's going on.

## Downgrading Qt

![image](https://github.com/qutebrowser/qutebrowser/assets/625793/39249bbf-0db8-470e-b550-7e8c1d6e679e)

A bit more scary, as it's less expected and more destructive. Note how the instructions how to change the backend aren't printed here, because the decision was already taken based on `QUTE_QT_WRAPPER`, so no need to tell the user what they already know.

## Version info

![image](https://github.com/qutebrowser/qutebrowser/assets/625793/e61e02cc-86c9-414b-ab73-a04d25db2d14)

![image](https://github.com/qutebrowser/qutebrowser/assets/625793/3e86edb5-dc75-443f-99d5-bbaadc384d27)

![image](https://github.com/qutebrowser/qutebrowser/assets/625793/c96756dd-13f4-44b1-873e-d38fb59b9065)


>> do we want to surface the wrapper selection logic (force or auto) in :version?

> yes

yes!

# Storing wrapper information in state file

>> is there any use case that'll be enabled by storing the last used wrapper in the state file? Eg if you have both 5 and 6 installed and forced Qt5 do we remember that and not do autoselection until you change to forcing Qt6?

> * I'm not sure. The use cases I can think of for this are a) make the wrapper selection "sticky" so you don't have to provide the env var every time; or b) provide a warning when the wrapper changes; or c) prevent downgrades via autoselection.
> * For (a) we currently only initialize the state file after reading all the config. So we can't use the state file to make it transparently sticky, we could use to to show a warning + restart-with-override option. But in the case where you would want the selection to be sticky that's probably not a great UX?
> * For (b) the wrapper changing is currently a one-to-one match to major Qt version changes which we already have a warning for. So it's redundant right now. Would it be useful in the future when we have multiple wrappers for the same Qt version? I don't know. If you had our preferred wrapper installed and working and we switched to it would it have any user relevant effects? Not enough data.
> * For (c) putting more barriers in place before a downgrade might be something we want to explore. But again, for now its the same case as a major version downgrade which we already have covered.
> * TL;DR: we can store it but I'm not sure what we would use it for.

Agreed. My initial thought with it was to avoid things getting "flaky" when a Qt install gets broken somehow. But we have a warning for this and also try to avoid any non-`ModuleNotFoundError` but `ImportError` and surface it instead of silently continuing. Let's not do that for now. It also has a big potential to make things more confusing because there is yet another (hidden) source of truth.

# More information in upgrade warnings

>> do we need to put more info in the upgrade warning like "because we detected a newer Qt version" or "because a different Qt version was selected by $mechanism"?
> * I guess it wouldn't hurt to put the wrapper override mechanism there if it was used. Although thinking about that opens up a can of worms in my lazy brain. Why not list all the available wrappers in there too? And tell people how to change it! And link to some docs! So much work.

Done. I also added a pointer to how to change things so users at least know where to start. Links are a bit tricky, because their browser will probably be qutebrowser, yet it's qutebrowser that currently refuses to start before they took a decision :sweat_smile: 


# Continue with other version button

> > in the upgrade warning, when you have both versions installed, should we add a "continue with $other_version" button?
>   
>   * We have this for the backend warnings but I'm not sure it makes sense here.
>   * For upgrades I guess this would be like "I don't want to upgrade just yet, ask me again next time". We could add it via restart + override but since there aren't any major issues with Qt6 and we are going to recommend people upgrade then I'm not sure it's worth the effort.
>   * For downgrades you should only be downgrading if you've specified the wrapper explicitly so I don't think it makes much sense there.

I don't think this is possible with how we do things right now: We use Qt to display the dialog, so (without storing something in the `state` file), we cannot switch anymore, and we cannot permanently switch and restart.

We could in theory restart with `--qt-wrapper PyQt[the-other-one]`, but yeah, probably not worth the effort, and I don't really want to encourage not upgrading.

# Other issues during import

> > is the current autoselection logic comprehensive enough? Currently we only try to import the top level packages (PyQt5/6), if something like importing WebEngine or QtNetwork later fails we wouldn't catch that till later. Is that something that is worth accounting for?

>  :shrug:  I have no opinion about this. It's too hypothetical for me.

Already ran into this during testing, because apparently `pip install PyQt6` then `pip uninstall PyQt6` weirdly results in `import PyQt6` still working, but e.g. `from PyQt6 import QtCore` breaking.

However, we already handle this in `earlyerror.py` - and I see this as one of those situations where we shouldn't just silently skip and use PyQt5, but complain about the setup being broken. This now pops up the `earlyinit.py` dialog but adds some information about the selected wrapper at the bottom:

![image](https://github.com/qutebrowser/qutebrowser/assets/625793/e7c47e21-2850-408f-8824-de9507f851ff)

The same error would also be displayed on any other unexpected `ImportError` exceptions (usually things like broken .so libraries after some upgrade without Qt rebuild).

Finally, if nothing at all is importable, that is also displayed by `earlyinit.py` correctly again now:

![image](https://github.com/qutebrowser/qutebrowser/assets/625793/cc43f051-a204-4cab-86c9-30036f09d10e)


> * can we add in a CLI argument for selecting the wrapper? Using an env var is a bit more cumbersome currently.
>   
>   * Yes we can, parsing the CLI args is the one thing we do before importing Qt.
>   * What would the arg be? `--qt-wrapper` I guess.

Done. And yes, `--qt-wrapper` is consistent with `QUTE_QT_WRAPPER`.

**Comment by ionenwks** [comment]
> ## Packagers overriding Qt version
> > > are any packagers expecting to force the Qt wrapper version?
> 
> > apparently so! I think it only makes sense when you are putting the choice on the user in an OS native way, like Gentoo's use flags apparently, or having multiple packages in Arch.
> 
> Decided to leave this out for the follow-up PR actually flipping the switch.
> 
> For Archlinux, I expect to merge the qutebrowser-qt6-git package into qutebrowser-git once the flip is switched and add PyQt6 as dependency. Someone in theory could upload a qutebrowser-qt5-git after that, but that's probably not going to be me unless there is actual demand for it.
> 
> This pretty much leaves us Gentoo. Unfortunately, the `_DEFAULT_WRAPPER = "PyQt5"` (as well as the given `sed` instructions) don't really make sense anymore after autodetection is on.
> 
> My proposal here when flipping the switch would be to change this to `_WRAPPER_OVERRIDE = None` which can be set by distributions (by patching the source), and then if that is set, use that instead of autodetection. I think `QUTE_QT_WRAPPER` and `--qt-wrapper` should still take precedence though.
> 
> Unfortunately this means distributions will need to adjust their patching. The naming/semantics of `_DEFAULT_WRAPPER` were a bit short-sighted in hindsight.
> 
> @ionenwks Any thoughts on this?

I see bit of patching adjustment and/or being non-intuitive as a non-issue. This will be temporary anyway, don't plan to keep the qt5 choice support around forever (already made qt6 the default and users need to manually opt-out if they don't want it yet, except on arches where we haven't tested qt6's webengine yet, like x86-32 and arm32 and it still use qt5 there for now).
